File: src/runtime/boxes/BbAny.cs
// BbAny.cs — ANY: match one char if in charset
// Mirrors src/runtime/boxes/bb_any.c
//
// α: if cursor >= Length or char not in charset → ω;  cursor++; γ(1 char)
// β: cursor--; ω

namespace Snobol4.Runtime.Boxes;

public sealed class BbAny : IByrdBox
{
    private readonly string _chars;
    public BbAny(string chars) { _chars = chars ?? ""; }

    public Spec Alpha(MatchState ms)
    {
        if (!ms.CharInSet(ms.Cursor, _chars)) return Spec.Fail;
        var result = Spec.Of(ms.Cursor, 1);
        ms.Cursor++;
        return result;
    }

    public Spec Beta(MatchState ms)
    {
        ms.Cursor--;
        return Spec.Fail;
    }
}

// BbNotany.cs — NOTANY: match one char if NOT in charset
// Mirrors src/runtime/boxes/bb_notany.c

public sealed class BbNotany : IByrdBox
{
    private readonly string _chars;
    public BbNotany(string chars) { _chars = chars ?? ""; }

    public Spec Alpha(MatchState ms)
    {
        if (ms.Cursor >= ms.Length)              return Spec.Fail;
        if (ms.CharInSet(ms.Cursor, _chars))     return Spec.Fail;
        var result = Spec.Of(ms.Cursor, 1);
        ms.Cursor++;
        return result;
    }

    public Spec Beta(MatchState ms)
    {
        ms.Cursor--;
        return Spec.Fail;
    }
}

File: src/runtime/boxes/BbAny.java
package snobol4.runtime.boxes;

/**
 * BbAny.java — ANY: match one char if in set
 * Port of bb_any.c / bb_any.s
 *
 *   ANY_α:  if (Δ>=Ω || chars.indexOf(Σ[Δ])<0)  goto ANY_ω;
 *           ANY=spec(Σ+Δ,1); Δ++;                goto ANY_γ;
 *   ANY_β:  Δ--;                                 goto ANY_ω;
 */
public class BbAny extends BbBox {
    private final String chars;

    public BbAny(MatchState ms, String chars) { super(ms); this.chars=chars; }

    @Override public Spec alpha() {
        if (ms.delta >= ms.omega) return null;
        if (chars.indexOf(ms.sigma.charAt(ms.delta)) < 0) return null;
        return new Spec(ms.delta++, 1);
    }

    @Override public Spec beta() { ms.delta--; return null; }
}
